opentracing-apache-httpclient/src/main/java/io/opentracing/contrib/apache/TracingHttpClientBuilder.java

```diff
@@ -0,0 +1,189 @@
+import io.opentracing.contrib.spanmanager.DefaultSpanManager;
+import io.opentracing.contrib.spanmanager.SpanManager;
+import io.opentracing.propagation.Format;
+import io.opentracing.tag.Tags;
+import io.opentracing.util.GlobalTracer;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.apache.http.HttpException;
+import org.apache.http.client.RedirectStrategy;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpExecutionAware;
+import org.apache.http.client.methods.HttpRequestWrapper;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.conn.routing.HttpRoute;
+import org.apache.http.impl.client.DefaultRedirectStrategy;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.execchain.ClientExecChain;
+
+/**
+ * @author Pavol Loffay
+ */
+public class TracingHttpClientBuilder extends org.apache.http.impl.client.HttpClientBuilder {
+    /**
+     * SpanContext which will be used as a parent for created client span.
+     */
+    public static final String PARENT_CONTEXT = TracingHttpClientBuilder.class.getName() + ".parentSpanContext";
+
+    /**
+     * Id of {@link HttpClientContext#setAttribute(String, Object)} representing span associated with
+     * the current request.
+     */
+    protected static final String ACTIVE_SPAN = TracingHttpClientBuilder.class.getName() + ".activeSpan";
+    /**
+     * Tracing {@link ClientExecChain} is executed after redirect exec, so on redirects it is called
+     * multiple times. This is used as an id for {@link HttpClientContext#setAttribute(String, Object)}
+     * to store number of redirects.
+     */
+    protected static final String REDIRECT_COUNT = TracingHttpClientBuilder.class.getName() + ".redirectCount";
+
+    private RedirectStrategy redirectStrategy;
+    private boolean redirectHandlingDisabled;
+
+    private Tracer tracer;
+    private SpanManager spanManager = DefaultSpanManager.getInstance();
+    private List<ApacheClientSpanDecorator> spanDecorators;
+
+    /**
+     * When using this constructor tracer should be registered via
+     * {@link GlobalTracer#register(Tracer)}.
+     */
+    public TracingHttpClientBuilder() {
+        this(DefaultRedirectStrategy.INSTANCE,
+                false,
+                GlobalTracer.get(),
+                Collections.<ApacheClientSpanDecorator>singletonList(
+                    new ApacheClientSpanDecorator.StandardTags()));
+    }
+
+    /**
+     * @param redirectStrategy redirect strategy, do not call
+     * {@link HttpClientBuilder#disableRedirectHandling()}
+     * @param redirectHandlingDisabled disable redirect strategy, do not call
+     * {@link org.apache.http.impl.client.HttpClientBuilder#setRedirectStrategy(RedirectStrategy)}
+     */
+    public TracingHttpClientBuilder(RedirectStrategy redirectStrategy,
```

Comment:
Why is this not using the builder pattern when called `Builder` - and setting up the items using `with...` methods?
Assuming something related to inheriting from HttpClientBuilder - if so, might be better to wrap instead of inherit to provide a cleaner builder?